backend/src/main/java/org/eclipse/tractusx/semantics/registry/service/ShellService.java

```diff
@@ -81,17 +96,63 @@ public Shell findShellByExternalId(String externalShellId) {
     }
 
     @Transactional(readOnly = true)
-    public ShellCollectionDto findAllShells(int page, int pageSize) {
-        Pageable pageable = PageRequest.of(page, pageSize, Sort.Direction.ASC, "createdDate");
-        Page<Shell> shellsPage = filterSpecificAssetIdsByTenantId(shellRepository.findAll(pageable));
-        return ShellCollectionDto.builder()
-                .currentPage(pageable.getPageNumber())
-                .totalItems((int) shellsPage.getTotalElements())
-                .totalPages(shellsPage.getTotalPages())
-                .itemCount(shellsPage.getNumberOfElements())
-                .items(shellsPage.getContent())
-                .build();
-    }
+    public ShellCollectionDto findAllShells(Integer pageSize,String cursorVal) {
+
+      pageSize = getPageSize( pageSize );
+      ShellCursor cursor = new ShellCursor( pageSize, cursorVal );
+      var specification = new ShellSpecification<Shell>( SORT_FIELD_NAME, cursor );
+
+      Page<Shell> shellPage = filterSpecificAssetIdsByTenantId( shellRepository.findAll( specification, ofSize( cursor.getRecordSize() ) ) );
+      var shellsPage = shellPage.getContent();
+
+      String nextCursor=null;
+
+      if(shellsPage.size()>0) {
+         nextCursor = cursor.getEncodedCursor(
+               shellsPage.get( shellsPage.size() - 1 ).getCreatedDate(),
+               shellPage.hasNext() );
+      }
+
+      return ShellCollectionDto.builder()
+            .items( shellsPage )
+            .cursor( nextCursor )
+            .build();
+   }
+
+   @Transactional( readOnly = true )
+   public SubmodelCollectionDto findAllSubmodel( Integer pageSize, String cursorVal, Shell assetID ) {
+      pageSize = getPageSize( pageSize );
+
+      ShellCursor cursor = new ShellCursor( pageSize, cursorVal );
+      var specification = new ShellSpecification<Submodel>( SORT_FIELD_NAME, cursor );
+      Page<Submodel> shellPage = submodelRepository.findAll( Specification.allOf( hasShellFkId( assetID ).and( specification ) ),
+            ofSize( cursor.getRecordSize() ) );
+
+      var shellsPage = shellPage.getContent();
+      String nextCursor=null;
+
+      if(shellsPage.size()>0) {
+         nextCursor = cursor.getEncodedCursor(
+               shellsPage.get( shellsPage.size() - 1 ).getCreatedDate(),
+               shellPage.hasNext()
+         );
+      }
+      return SubmodelCollectionDto.builder()
+            .items( shellsPage )
+            .cursor( nextCursor )
+            .build();
+   }
+
+   private static Integer getPageSize( Integer pageSize ) {
+      if ( pageSize == null ) {
+         pageSize = Integer.MAX_VALUE;
```

Comment:
Max Value is 2147483647? Is there a limitation or to we try to get at least 2147483647 entries from the database.
Maybe it would make sense to define our own max value.

Reply:
Actually the default page size in spring data is 1000 or 2000. Only if we increase it explicitly, then the number of records will increase. source: https://medium.com/@rafiek/spring-data-pageable-maximum-page-size-1bcc9f5348aa

Reason behind giving this Integer.MAX_VALUE is to provide the maximum upper value from our side, so that Spring JPA will automatically handle it and provide the maximum default size resultset.
But yes I don't mind giving any upper limit also. I can specify it with next change if that is fine for you.